### PR TITLE
Improvement for ProtobufTranslator

### PR DESCRIPTION
**Pull request by oliviertoupin**
- Fix N^2 in encoding (linear search of a map)
- IdentifyHashMap to make sure there is no regression or problem with redifined hashcode
- Comments on the implementation
- Formatting.

### CODE REVIEW

**Comment by oliviertoupin** [comment]
@alexb @coderunner

File: mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala [hunk 1]
```diff
@@ -60,16 +61,32 @@ class ProtobufTranslator extends ProtocolTranslator {
...
   private var currentHeapId = 1 // I don't think will have more than 2^31 object, skip 0, since 0==unassigned
 
   private val pb2obj = new collection.mutable.HashMap[Int, AnyRef] // Encoded HeapId to decoded instance mapping
-  private val obj2pb = new collection.mutable.HashMap[AnyRef, Int] // Live instance to encoded HeapId mapping
+
+  private val obj2pb = new util.IdentityHashMap[AnyRef, Int] // Live instance to encoded HeapId mapping.
 
   // Used by encoding and decoding as temporary storage for encoded/decoded object
-  private var tempHeap = new collection.mutable.ListBuffer[AnyRef]
+  private var tempHeap = new collection.mutable.ArrayBuffer[AnyRef]
```
Comment (coderunner) [inline mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala]: Why a var here?

File: mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala [hunk 2]
```diff
@@ -298,7 +314,12 @@ private class InternalProtobufTranslator {
     varId.zipWithIndex.foreach((v) => registerEncodedData(v._1, block.variables(v._2)))
     pBlock.addAllVariableHeapIds(varId.map(_.asInstanceOf[java.lang.Integer]))
 
-    pBlock.addAllOperationHeapIds(block.operations.map(encodePOperation(_)).map(addToHeap(_)).map(_.asInstanceOf[java.lang.Integer]))
+    def addOperation(op: Operation) = {
+      val pOp = encodePOperation(op)
+      addToHeap(pOp).asInstanceOf[java.lang.Integer]
+    }
```
Comment (oliviertoupin) [inline mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala:319]: This is pertinent

File: mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala [hunk 3]
```diff
@@ -298,7 +314,12 @@ private class InternalProtobufTranslator {
     varId.zipWithIndex.foreach((v) => registerEncodedData(v._1, block.variables(v._2)))
     pBlock.addAllVariableHeapIds(varId.map(_.asInstanceOf[java.lang.Integer]))
 
-    pBlock.addAllOperationHeapIds(block.operations.map(encodePOperation(_)).map(addToHeap(_)).map(_.asInstanceOf[java.lang.Integer]))
```
Comment (coderunner) [inline mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala:301]: good delete! :)

File: mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala (hunk 1 above)
Comment (oliviertoupin) [inline mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala]: No reason

File: mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala [hunk 4]
```diff
@@ -60,16 +61,32 @@ class ProtobufTranslator extends ProtocolTranslator {
...
+  // would save that copying in some cases. All of this is linear time however, so it's an optimization, not a priority.
+  // However, there is a few transaction, if any, of that size and it's still manageable, it just not optimal.
+  // Also, the current implementation is probably more readable.
+
   private var currentHeapId = 1 // I don't think will have more than 2^31 object, skip 0, since 0==unassigned
 
   private val pb2obj = new collection.mutable.HashMap[Int, AnyRef] // Encoded HeapId to decoded instance mapping
-  private val obj2pb = new collection.mutable.HashMap[AnyRef, Int] // Live instance to encoded HeapId mapping
+
+  private val obj2pb = new util.IdentityHashMap[AnyRef, Int] // Live instance to encoded HeapId mapping.
```
Comment (oliviertoupin) [inline mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala:85]: IdentityHashMap

File: mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala [hunk 5]
```diff
@@ -60,16 +61,32 @@ class ProtobufTranslator extends ProtocolTranslator {
  * Allow stateful and threadsafe operation for ProtobufTranslator by instancing this class at every call.
  * It also keep ProtobufTranslator class more readable and more relevant to it's parent interface.
  *
+ * Note on implementation: The message is structured somewhat like a computer heap. All the data is stored without so
```
Comment (oliviertoupin) [inline mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala:64]: Comment that should have been there all along.

File: mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala [hunk 6]
```diff
@@ -60,16 +61,32 @@ class ProtobufTranslator extends ProtocolTranslator {
...
  */
 private class InternalProtobufTranslator {
 
+
+  // TODO: This implementation is inefficient (while good enough) for large number of operation (ex: n >= 10000).
+  // It copies everything from the original Transaction to temporary lists, and do a lot of map, zip and iteration over
+  // it, and then copies it again to the ProtocolBuffer structure. Using iterator instead of temporary lists,
+  // would save that copying in some cases. All of this is linear time however, so it's an optimization, not a priority.
+  // However, there is a few transaction, if any, of that size and it's still manageable, it just not optimal.
+  // Also, the current implementation is probably more readable.
```
Comment (oliviertoupin) [inline mry-core/src/main/scala/com/wajam/mry/api/protobuf/ProtobufTranslator.scala:79]: Future improvements?

**Comment by coderunner** [comment]
+1

**Pull request merged**